### PR TITLE
test(bridge-flows): characterization tests for page.tsx

### PR DESCRIPTION
**Pull request by chapati23**
## Summary

- Adds 26 characterization tests at `ui-dashboard/src/app/bridge-flows/__tests__/page.test.tsx` (1556 → 1582 in the suite total). Pure test addition — zero production code changes.
- Pins the page's current behavior across **7 categories**: initial render (5), TransfersTable sort (5), pagination clamping (1, with offset capture), status filter (4), toast lifecycle (5), RouteDeliveryTile (3), Hasura-not-configured friendly error (1).
- Establishes the **TDD safety net** for the next two PRs in the series: A2 (extract `_components/transfers-table.tsx` + row cells) and A3 (extract `_components/route-delivery-tile.tsx` + chart-section).

## Approach

- Mocks: `useBridgeGQL` (multi-query discriminator driven by GraphQL query string), `useOracleRates`, `next/navigation`, `@/components/address-labels-provider`, three chart components (data-testid markers only — plotly is intentionally not exercised), `BridgeRedeemPill`/`ToastPortal`, `AddressLink`, `ChainIcon`.
- Captured-addToast tests use `vi.doMock` + `vi.resetModules` per-test, scoped to one describe block with explicit `afterEach` teardown.
- Pattern mirrors `app/pool/[poolId]/__tests__/page.test.tsx`.
- Fixtures use proper 40-char `0x[a-f0-9]{40}` addresses + 64-char tx hashes via top-level `NEW_SENDER`/`MID_SENDER`/`OLD_SENDER` constants — any future production-side hex validation can't pass spuriously on these fixtures.

## Out-of-scope (deferred)

- Real SSR coverage. The `renderToStaticMarkup` smoke tests originally bundled here ran in `@vitest-environment jsdom` so `window` was available, defeating their purpose. If A2/A3 wants SSR coverage it should land as a sibling `page.ssr.test.tsx` with `@vitest-environment node`.

## Test plan

- [x] `pnpm --filter @mento-protocol/ui-dashboard test` — 1582 pass (1556 baseline + 26 new)
- [x] `pnpm --filter @mento-protocol/ui-dashboard typecheck` — clean
- [x] `./tools/trunk fmt && ./tools/trunk check` — clean
- [ ] CI: confirm Vercel preview build succeeds

🤖 Generated with [Claude Code](https://claude.com/claude-code)

---

> [!NOTE]
> **Low Risk**
> Low risk because the PR only adds tests and does not modify production logic. Main impact is increased test runtime/maintenance due to extensive mocking and behavior pinning.
> 
> **Overview**
> Adds a large `vitest` jsdom characterization suite for `app/bridge-flows/page.tsx`, with mocks for `useBridgeGQL`, `next/navigation`, oracle rates, and heavy UI components.
> 
> The tests pin behavior across initial rendering, **per-query loading states**, table sorting (including numeric route ordering and null-sink semantics), pagination clamping and URL updates via `router.replace`, status-filter URL/state wiring, toast enqueue/auto-dismiss behavior, route-delivery tile aggregation/exclusion rules, and the friendly "Hasura not configured" error banner path.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit b936c6802924eb59ffb2522026a7c04f7fb7a412. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

**Comment by chapati23** [comment]
@claude review

**Comment by chapati23** [comment]
cursor review

File: ui-dashboard/src/app/bridge-flows/__tests__/page.test.tsx [hunk 1]
```diff
@@ -0,0 +1,1108 @@
...
+// Fixtures
+// ---------------------------------------------------------------------------
+
+type GQLResult<T> = {
+  data: T | undefined;
+  error: Error | undefined;
+  isLoading: boolean;
+};
+
+function ok<T>(data: T): GQLResult<T> {
```
[the remark this answers is not in the document]
Comment (chapati23) [reply ui-dashboard/src/app/bridge-flows/__tests__/page.test.tsx:187]: Fixed in 2f593aa: added a 5-test `loading matrix` describe block. Each test isolates `isLoading: true` on a single query slot (transfers, snapshots, topBridgers, pending, delivered) and asserts the corresponding loading UI: `<Skeleton role="status">` for transfers, `data-loading="true"` propagated to chart stubs (BridgeVolumeChart / BridgeTokenBreakdownChart / BridgeTopBridgersChart), "…" placeholder in the In-Flight tile, the 2-bar `.animate-pulse` skeleton in RouteDeliveryTile. Chart stubs were also updated to surface `isLoading` via `data-loading` so the assertion is meaningful (no production code changes).

File: ui-dashboard/src/app/bridge-flows/__tests__/page.test.tsx [hunk 2]
```diff
@@ -0,0 +1,1108 @@
...
+    // 3 data rows in tbody.
+    expect(container.querySelectorAll("tbody tr")).toHaveLength(3);
+  });
+});
+
+// ---------------------------------------------------------------------------
+// 2. TransfersTable sort state
+// ---------------------------------------------------------------------------
+
+describe("BridgeFlowsPage — TransfersTable sort state", () => {
```
[the remark this answers is not in the document]
Comment (chapati23) [reply ui-dashboard/src/app/bridge-flows/__tests__/page.test.tsx:561]: Fixed in 2f593aa: added 3 direct tests for the previously-uncovered comparators in the page-level sort block. `route` test mixes MONAD->CELO and CELO->MONAD fixtures and asserts numeric ASC/DESC ordering on the chainId pair (not lexical — so a regression to lex compare would flip the order). `amountUsd` and `duration` tests both include null-valued rows and assert null sinks to the bottom on **both** ASC and DESC sort directions (the contract is null-sinks-regardless-of-direction). Note: lib-level tests in `lib/bridge-flows/__tests__/sort.test.ts` already cover `sortTransfers` directly; the new tests duplicate that contract through the page wiring so an extraction that miswires a header to the wrong key still fails the suite.

File: ui-dashboard/src/app/bridge-flows/__tests__/page.test.tsx [hunk 3]
```diff
@@ -0,0 +1,1108 @@
...
+    expect(order[1]).toContain(MID_SENDER);
+    expect(order[2]).toContain(NEW_SENDER);
+  });
+});
+
+// ---------------------------------------------------------------------------
+// 3. Pagination clamping
+// ---------------------------------------------------------------------------
+
+describe("BridgeFlowsPage — pagination clamping", () => {
```
[the remark this answers is not in the document]
Comment (chapati23) [reply ui-dashboard/src/app/bridge-flows/__tests__/page.test.tsx]: Fixed in 2f593aa: added 5 tests in a new `outgoing setPage -> router.replace` sub-block that click the Pagination component buttons (`aria-label="Next page"`, `Last page`, `First page`, `Previous page`) and assert the URL passed to `mockReplace`:

- Next from page 1 -> URL contains `page=2`
- Last from page 1 of 3 -> URL contains `page=3`
- First from page 2 -> URL has NO `page=` token (page=1 is canonicalized as URL-cleared, matching the `params.delete("page")` call in `setPage`)
- Prev from page 2 -> same NO `page=` canonicalization
- Page navigation with `?status=PENDING` URL preserves `status=PENDING` while advancing the page (`status` and `page` are independent URL-state slots)

File: ui-dashboard/src/app/bridge-flows/__tests__/page.test.tsx [hunk 4]
```diff
@@ -0,0 +1,1108 @@
...
+      }),
+    );
+    renderJsdom();
+    // The route delivery tile section title should still appear, with an
+    // em-dash (—) as the body.
+    expect(container.textContent).toContain("Avg Delivery Time by Route");
+    expect(container.textContent).toContain("—");
+  });
+
+  it("renders per-route averages when delivered transfers are returned", () => {
```
[the remark this answers is not in the document]
Comment (chapati23) [reply ui-dashboard/src/app/bridge-flows/__tests__/page.test.tsx:1509]: Fixed in 2f593aa: added 4 tests covering the `computeRouteAvgDeliverTimes` exclusion contract through the page (route-stats.test.ts already pins the lib-level shape; these tests pin it through the page wiring so a future change to the BRIDGE_DELIVERED_RECENT query slot can't quietly start summing partial rows):

- non-DELIVERED status with both timestamps set (SENT, ATTESTED) — must NOT pollute the average; n=1 and the `16m 40s` formatted average pin the contract
- null `sourceChainId` or null `destChainId` — both excluded; only one `n=N` line renders
- null `sentTimestamp` or null `deliveredTimestamp` — duration unknown, both shapes excluded
- mixed shape (one valid + one of every excluded class) — n=1 and the formatted average prove only the valid sample contributed

**Comment by chapati23** [comment]
All 4 review findings addressed in 2f593aa with +17 tests (25 -> 42, all passing).

| Finding | Tests added |
|---|---|
| Loading matrix unpinned | +5 — isolate `isLoading: true` on each query slot (transfers, snapshots, topBridgers, pending, delivered) and assert the right loading UI renders |
| `route` / `amountUsd` / `duration` comparators not covered by page-level sort tests | +3 — direct tests with null-valued fixtures asserting null-sink semantics on both ASC and DESC |
| Outgoing `setPage -> router.replace` URL-replace branch untested | +5 — click Next/Last/First/Prev, assert URL contains the right `page=N`, `page=1` is canonicalized to NO `page=` token, `?status=` survives navigation |
| RouteDeliveryTile only covered clean DELIVERED rows | +4 — non-DELIVERED with both timestamps, null chain IDs, null timestamps, mixed shape — all excluded from the average |

Chart stubs were updated to surface `isLoading` via `data-loading` so the loading-matrix assertion is meaningful (no production code changes; only the test harness was touched). `pnpm typecheck`, `pnpm vitest run`, and `./tools/trunk fmt + check` all pass.

**Comment by chapati23** [comment]
@claude — addressed in 416eecf. Summary:

- **`rowOrder()` fragility** → resolves the Sender column by header text now, walking `thead th` and finding the one whose textContent includes 'Sender'. If A2 reorders columns, the assertion stays correct. If 'Sender' is removed, we fail loudly with `'Sender' header not found in thead` instead of returning data from the wrong column.
- **Describe block name** → renamed "pagination clamping" to "pagination and count state" so the count-error fallback test no longer feels misplaced.
- **'In progress' test name** → renamed to use the actual production copy ('No bridge transfers match the selected status').
- **Broad textContent assertions** → tightened initial-render to `container.querySelector('h1')?.textContent` for the heading and a `paragraphs.some(...)` lookup for the subtitle; can no longer pass on a tooltip / error banner that happens to contain the searched string.

Skipped (non-blocking style points):
- **stuck fixture duplicated 3×** — each of the 3 toast tests uses distinct ids/senders/recipients deliberately; a shared `makeStuckTransfer(seed)` would be a small win but every line is currently load-bearing for the assertion it supports.
- **Lone toast-smoke describe** — kept as-is. The smoke-vs-captured-via-doMock boundary is the whole point of having two describes; collapsing them would make it less clear which test needs the per-test `vi.doMock` scope.

All 42 tests still green. Thanks for the careful read.

File: ui-dashboard/src/app/bridge-flows/__tests__/page.test.tsx [hunk 5]
```diff
@@ -0,0 +1,1757 @@
...
+    const r2 = makeTransfer({
+      id: "r2",
+      sender: MID_SENDER,
+      sourceChainId: 42220, // CELO
+      destChainId: 143, // -> MONAD
+    });
+    mockUseBridgeGQL.mockImplementation(
+      bridgeImpl({
+        transfers: ok({ BridgeTransfer: [r2, r1] }),
+        count: ok({ BridgeTransfer: [{ id: r1.id }, { id: r2.id }] }),
```
[the remark this answers is not in the document]
Comment (chapati23) [reply ui-dashboard/src/app/bridge-flows/__tests__/page.test.tsx]: Fixed in 236e523. Added a third row with `sourceChainId: 8453` so the numeric and lexical orderings now diverge: numeric ASC = [143, 8453, 42220], lexical ASC = [143, 42220, 8453]. The middle-row assertion in ASC and the first-row assertion in DESC both flip under a numeric→lexical regression, so the test fails loudly on the bug class you flagged. Thanks for the catch.

[the remark this answers is not in the document]
Comment (chapati23) [reply ui-dashboard/src/app/bridge-flows/__tests__/page.test.tsx]: Fixed in b936c68. Added a third row with `sourceChainId: 8453` so the numeric and lexical orderings now diverge: numeric ASC = [143, 8453, 42220], lexical ASC = [143, 42220, 8453]. The middle-row assertion in ASC and the first-row assertion in DESC both flip under a numeric→lexical regression, so the test fails loudly on the bug class you flagged. Thanks for the catch.
